File: backend/app/schemas/memo.py
import logging
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, HttpUrl
from app.models.memo import MemoStatus

logger = logging.getLogger(__name__)


# ── 요청 스키마 ────────────────────────────────────────────────

class MemoCreate(BaseModel):
    """메모 생성 요청"""
    content: str
    source_url: Optional[str] = None


class MemoStatusUpdate(BaseModel):
    """상태 변경 요청 (칸반 드래그 등)"""
    status: MemoStatus


class ParseUrlRequest(BaseModel):
    """URL 메타데이터 파싱 요청"""
    url: str


# ── 응답 스키마 ────────────────────────────────────────────────

class DialogueExchange(BaseModel):
    """대화문 한 줄"""
    speaker: str   # "A" | "B"
    line: str
    korean: str


class AIDialogue(BaseModel):
    """AI 생성 대화문 전체 구조"""
    title: str
    situation: str
    exchanges: List[DialogueExchange]


class MemoResponse(BaseModel):
    """메모 단건 응답"""
    id: int
    user_id: str
    content: str
    source_url: Optional[str] = None
    url_title: Optional[str] = None
    url_description: Optional[str] = None
    status: MemoStatus
    start_date: datetime
    end_date: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    is_transformed: bool
    ai_summary_ko: Optional[str] = None
    ai_summary_en: Optional[str] = None
    ai_dialogue_json: Optional[str] = None   # JSON 문자열 그대로 전달
    audio_s3_key: Optional[str] = None

    model_config = {"from_attributes": True}


class TransformResponse(BaseModel):
    """AI 변환 결과 응답"""
    summary_ko: str
    summary_en: str
    dialogue: AIDialogue
    credits_remaining: int


class CreditsResponse(BaseModel):
    """크레딧 현황 응답"""
    daily_credits: int
    is_premium: bool
    max_daily_credits: int

File: frontend/src/types/memo.ts
export type MemoStatus = 'not_started' | 'in_progress' | 'completed' | 'keep_reviewing';

export const STATUS_LABEL: Record<MemoStatus, string> = {
  not_started: '진행 전',
  in_progress: '진행 중',
  completed: '완료',
  keep_reviewing: '계속 참조',
};

export const STATUS_COLOR: Record<MemoStatus, string> = {
  not_started: 'bg-red-100 text-red-600',
  in_progress: 'bg-yellow-100 text-yellow-700',
  completed: 'bg-green-100 text-green-700',
  keep_reviewing: 'bg-blue-100 text-blue-700',
};

export interface DialogueExchange {
  speaker: 'A' | 'B';
  line: string;
  korean: string;
}

export interface AIDialogue {
  title: string;
  situation: string;
  exchanges: DialogueExchange[];
}

export interface Memo {
  id: number;
  user_id: string;
  content: string;
  source_url?: string;
  url_title?: string;
  url_description?: string;
  status: MemoStatus;
  start_date: string;
  end_date?: string;
  created_at: string;
  updated_at: string;
  is_transformed: boolean;
  ai_summary_ko?: string;
  ai_summary_en?: string;
  ai_dialogue_json?: string;
  audio_s3_key?: string;
}

export interface TransformResult {
  summary_ko: string;
  summary_en: string;
  dialogue: AIDialogue;
  credits_remaining: number;
}

export interface Credits {
  daily_credits: number;
  is_premium: boolean;
  max_daily_credits: number;
}
